# Detailed comparison between deep and non-deep learning

## By Brecht De Boeck - on behalf of Group 10

The difference between deep learning and non-deep learning is very simple: Any neural network with at least three layers is considered deep learning.

For this part of our assignment, I'll be creating two separate image classification models, one deep learning and one non-deep, and comparing the two based on both the speed at which they make a prediction, and their overall performance.

### Creating the non-deep learning models

In [123]:
import numpy as np
import pandas as pd
import cv2 as cv
import tensorflow as tf
import matplotlib.pyplot as plt
import os

In [124]:
#Read the basic data from the labeled dataset
data = pd.read_csv("data/swarmIoT_full_labeled_dataset.csv")
images = data["image"].str.split("-", n = 1, expand = True)
data["image"] = images[1]

Neural networks are the most obvious choice when it comes to image classification, but these generally fall under deep learning. We could try using a neural network without any hidden layers, where input and output are directly connected.

However, we have seen a lot of non-deep learning models throughout this course. Sure, none of them were explicitly used for image classification, but that doesn't mean we can't try using one of them for it. But then the question remains: which algorithm do we choose?

Our options:
- kNN: Nope. Not using this one. Not only is it way too slow, but it's pretty much undeployable, seeing as there isn't a real model that's created.
- Support Vector Machines: While an interesting choice, these kinds of models don't do very well in the way of speed and scaleability. Especially in the case of that first one, we don't want to give our neural network an unfair advantage.
- Decision Trees: Seems to be our best fit. Though it cannot actually see the images, but only their raw values, it is capable of making decisions based on the given input, similar to how a deep learning neural network would do it.
- Logistic Regression: As bizarre as it might seem, we can also use Logistic Regression if we reshape the dimensions of our training and test set to consist of at most two dimensions.

For this task, I'll be making both a Decision Tree and a Logistic Regression model. The Decision Tree (at least from my perspective) seems to lie the closest to emulating a neural network, while for Logistic Regression, I'm simply curious to find out what results the model might give us.

And finally, as discussed, I'll be making a neural network without any hidden layers. All of these models will be compared to our deep learning models made in the next chapter of this notebook.

#### Extracting features from the images

The problem with decision trees is that, of course, we can't just feed raw images to our model. The decision tree expects features with differing values, which it can then use to make splits to gain as much information gain as possible. The same goes for our Logistic Regression model: we can't just expect to be able to throw in raw image data and have the model make accurate predictions. Our first order of business is thus to create a new dataset, that combines our classifiers and their images with features we extract from the images.

In [125]:
from sklearn.cluster import MiniBatchKMeans
import time
import os

In [146]:
# The following function will be used to process every image within our dataset
def extract_features_from_image(camera, partition, img):
    st = time.time()
    path = "/".join(['data','images',camera,partition,img])
    partition_path = "/".join(['data','images',camera,partition])
    image = plt.imread(path)
    
    dicto = []
    
    if image is not None:
        #Extract features using SIFT
        sift = cv.SIFT_create()
        kp, des = sift.detectAndCompute(image,None)
        for d in des:
            dicto.append(d)
        
        k = 30
        batch_size = np.size(os.listdir(partition_path)) * 3
        kmeans = MiniBatchKMeans(n_clusters=k, batch_size=batch_size, verbose=1).fit(dico)
        
        histo = np.zeros(k)
        nkp = np.size(kp)
        for d in des:
            idx = kmeans.predict([d.astype(np.double)])
            histo[idx] += 1/nkp
    return histo

Using the above method should work when working with machine learning models, but we cannot use it on all of our images. That would simply be a fool's errand, clustering the keypoints of more than 6000 images. It's a good thing our images are split amongst partitions.

I'll be using the images from Camera 3, specifically the first partition, which our group labeled ourselves and only contains 856 images. I'll be taking the first 50 images from this partition and splitting it into training and test datasets. That's all the abuse I'm willing to inflict on my poor computer, even if the resulting sample won't give us nearly as much information as when working on the full dataset or even just a simple partition.

**For the Record**: Yes, I did try getting the features of an entire partition. That attempt resulted in my computer crashing. Again. I work with what I have in this assignment, so please just bare with me.

In [148]:
sample = os.listdir("data/images/3 De nayer (Loop piste)/partition 1")
sample = data.loc[data["image"].isin(sample)].head(50)

#extract_features_from_image("3 De nayer (Loop piste)", "partition 1",sample["image"].head(1).iloc[0])
sample_feats = [extract_features_from_image("3 De nayer (Loop piste)", "partition 1", img) for img in sample["image"]]

C:\Users\Brech\anaconda3\envs\chidi\lib\site-packages\sklearn\cluster\_kmeans.py:1043: UserWarning: MiniBatchKMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can prevent it by setting batch_size >= 3072 or by setting the environment variable OMP_NUM_THREADS=9
  warnings.warn(


Init 1/3 with method k-means++
Inertia for init 1/3: 291826708.0
Init 2/3 with method k-means++
Inertia for init 2/3: 290078818.0
Init 3/3 with method k-means++
Inertia for init 3/3: 286772473.0
Minibatch step 1/100: mean batch inertia: 137716.42173288655
Minibatch step 2/100: mean batch inertia: 91830.39866856401, ewa inertia: 91830.39866856401
Minibatch step 3/100: mean batch inertia: 89250.01682988797, ewa inertia: 89250.01682988797
Minibatch step 4/100: mean batch inertia: 89442.3004474216, ewa inertia: 89442.3004474216
Minibatch step 5/100: mean batch inertia: 89046.21283550088, ewa inertia: 89046.21283550088
Minibatch step 6/100: mean batch inertia: 89473.04884279845, ewa inertia: 89473.04884279845
Minibatch step 7/100: mean batch inertia: 88998.75502006288, ewa inertia: 88998.75502006288
Minibatch step 8/100: mean batch inertia: 87933.95818179616, ewa inertia: 87933.95818179616
Minibatch step 9/100: mean batch inertia: 88829.28346020142, ewa inertia: 88829.28346020142
Minibatch 

C:\Users\Brech\anaconda3\envs\chidi\lib\site-packages\sklearn\cluster\_kmeans.py:1043: UserWarning: MiniBatchKMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can prevent it by setting batch_size >= 3072 or by setting the environment variable OMP_NUM_THREADS=9
  warnings.warn(


Init 1/3 with method k-means++
Inertia for init 1/3: 289068101.0
Init 2/3 with method k-means++
Inertia for init 2/3: 286070393.0
Init 3/3 with method k-means++
Inertia for init 3/3: 288896724.0
Minibatch step 1/100: mean batch inertia: 137437.56199138344
Minibatch step 2/100: mean batch inertia: 93037.85649165805, ewa inertia: 93037.85649165805
Minibatch step 3/100: mean batch inertia: 90028.59801615696, ewa inertia: 90028.59801615696
Minibatch step 4/100: mean batch inertia: 89278.1878997089, ewa inertia: 89278.1878997089
Minibatch step 5/100: mean batch inertia: 89844.75701503576, ewa inertia: 89844.75701503576
Minibatch step 6/100: mean batch inertia: 89119.92515245198, ewa inertia: 89119.92515245198
Minibatch step 7/100: mean batch inertia: 89154.16215770101, ewa inertia: 89154.16215770101
Minibatch step 8/100: mean batch inertia: 89421.00648220864, ewa inertia: 89421.00648220864
Minibatch step 9/100: mean batch inertia: 88220.81565188276, ewa inertia: 88220.81565188276
Minibatch 

C:\Users\Brech\anaconda3\envs\chidi\lib\site-packages\sklearn\cluster\_kmeans.py:1043: UserWarning: MiniBatchKMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can prevent it by setting batch_size >= 3072 or by setting the environment variable OMP_NUM_THREADS=9
  warnings.warn(


Init 1/3 with method k-means++
Inertia for init 1/3: 288073303.0
Init 2/3 with method k-means++
Inertia for init 2/3: 287760085.0
Init 3/3 with method k-means++
Inertia for init 3/3: 287575485.0
Minibatch step 1/100: mean batch inertia: 138234.52369554812
Minibatch step 2/100: mean batch inertia: 92068.73041253048, ewa inertia: 92068.73041253048
Minibatch step 3/100: mean batch inertia: 90277.25502564685, ewa inertia: 90277.25502564685
Minibatch step 4/100: mean batch inertia: 89443.67300733752, ewa inertia: 89443.67300733752
Minibatch step 5/100: mean batch inertia: 89276.95389395955, ewa inertia: 89276.95389395955
Minibatch step 6/100: mean batch inertia: 88712.47654428866, ewa inertia: 88712.47654428866
Minibatch step 7/100: mean batch inertia: 88268.32289911073, ewa inertia: 88268.32289911073
Minibatch step 8/100: mean batch inertia: 88842.7029254882, ewa inertia: 88842.7029254882
Minibatch step 9/100: mean batch inertia: 88568.6714196585, ewa inertia: 88568.6714196585
Minibatch st

C:\Users\Brech\anaconda3\envs\chidi\lib\site-packages\sklearn\cluster\_kmeans.py:1043: UserWarning: MiniBatchKMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can prevent it by setting batch_size >= 3072 or by setting the environment variable OMP_NUM_THREADS=9
  warnings.warn(


Init 1/3 with method k-means++
Inertia for init 1/3: 291589989.0
Init 2/3 with method k-means++
Inertia for init 2/3: 289753095.0
Init 3/3 with method k-means++
Inertia for init 3/3: 290657245.0
Minibatch step 1/100: mean batch inertia: 137497.52321685015
Minibatch step 2/100: mean batch inertia: 91146.25572444827, ewa inertia: 91146.25572444827
Minibatch step 3/100: mean batch inertia: 89922.56653992823, ewa inertia: 89922.56653992823
Minibatch step 4/100: mean batch inertia: 89720.82157424149, ewa inertia: 89720.82157424149
Minibatch step 5/100: mean batch inertia: 88952.55425015138, ewa inertia: 88952.55425015138
Minibatch step 6/100: mean batch inertia: 88819.58304383726, ewa inertia: 88819.58304383726
Minibatch step 7/100: mean batch inertia: 89330.93698041106, ewa inertia: 89330.93698041106
Minibatch step 8/100: mean batch inertia: 89149.73899283912, ewa inertia: 89149.73899283912
Minibatch step 9/100: mean batch inertia: 88199.99507673422, ewa inertia: 88199.99507673422
Minibatc

C:\Users\Brech\anaconda3\envs\chidi\lib\site-packages\sklearn\cluster\_kmeans.py:1043: UserWarning: MiniBatchKMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can prevent it by setting batch_size >= 3072 or by setting the environment variable OMP_NUM_THREADS=9
  warnings.warn(


Init 1/3 with method k-means++
Inertia for init 1/3: 282317265.0
Init 2/3 with method k-means++
Inertia for init 2/3: 284151586.0
Init 3/3 with method k-means++
Inertia for init 3/3: 283665220.0
Minibatch step 1/100: mean batch inertia: 135738.3671613212
Minibatch step 2/100: mean batch inertia: 92126.75250937606, ewa inertia: 92126.75250937606
Minibatch step 3/100: mean batch inertia: 88572.05091407632, ewa inertia: 88572.05091407632
Minibatch step 4/100: mean batch inertia: 89117.69563691698, ewa inertia: 89117.69563691698
Minibatch step 5/100: mean batch inertia: 89681.59760618818, ewa inertia: 89681.59760618818
Minibatch step 6/100: mean batch inertia: 88859.17861693642, ewa inertia: 88859.17861693642
Minibatch step 7/100: mean batch inertia: 87662.15855759766, ewa inertia: 87662.15855759766
Minibatch step 8/100: mean batch inertia: 88097.38044701224, ewa inertia: 88097.38044701224
Minibatch step 9/100: mean batch inertia: 88170.11760439507, ewa inertia: 88170.11760439507
Minibatch

C:\Users\Brech\anaconda3\envs\chidi\lib\site-packages\sklearn\cluster\_kmeans.py:1043: UserWarning: MiniBatchKMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can prevent it by setting batch_size >= 3072 or by setting the environment variable OMP_NUM_THREADS=9
  warnings.warn(


Init 1/3 with method k-means++
Inertia for init 1/3: 285830062.0
Init 2/3 with method k-means++
Inertia for init 2/3: 287572906.0
Init 3/3 with method k-means++
Inertia for init 3/3: 290911700.0
Minibatch step 1/100: mean batch inertia: 135246.11010052657
Minibatch step 2/100: mean batch inertia: 92453.65364220366, ewa inertia: 92453.65364220366
Minibatch step 3/100: mean batch inertia: 91192.20409500744, ewa inertia: 91192.20409500744
Minibatch step 4/100: mean batch inertia: 89457.4150296668, ewa inertia: 89457.4150296668
Minibatch step 5/100: mean batch inertia: 89318.54344934147, ewa inertia: 89318.54344934147
Minibatch step 6/100: mean batch inertia: 88331.97974098456, ewa inertia: 88331.97974098456
Minibatch step 7/100: mean batch inertia: 88643.30806831195, ewa inertia: 88643.30806831195
Minibatch step 8/100: mean batch inertia: 88561.21812641785, ewa inertia: 88561.21812641785
Minibatch step 9/100: mean batch inertia: 87674.91435437408, ewa inertia: 87674.91435437408
Minibatch 

C:\Users\Brech\anaconda3\envs\chidi\lib\site-packages\sklearn\cluster\_kmeans.py:1043: UserWarning: MiniBatchKMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can prevent it by setting batch_size >= 3072 or by setting the environment variable OMP_NUM_THREADS=9
  warnings.warn(


Init 1/3 with method k-means++
Inertia for init 1/3: 286797720.0
Init 2/3 with method k-means++
Inertia for init 2/3: 288835677.0
Init 3/3 with method k-means++
Inertia for init 3/3: 290040655.0
Minibatch step 1/100: mean batch inertia: 136433.2647199617
Minibatch step 2/100: mean batch inertia: 90421.23927364034, ewa inertia: 90421.23927364034
Minibatch step 3/100: mean batch inertia: 90680.7840392156, ewa inertia: 90680.7840392156
Minibatch step 4/100: mean batch inertia: 90364.69651061854, ewa inertia: 90364.69651061854
Minibatch step 5/100: mean batch inertia: 88631.74955727976, ewa inertia: 88631.74955727976
Minibatch step 6/100: mean batch inertia: 88353.96435248062, ewa inertia: 88353.96435248062
Minibatch step 7/100: mean batch inertia: 89019.85765884095, ewa inertia: 89019.85765884095
Minibatch step 8/100: mean batch inertia: 88184.59908227212, ewa inertia: 88184.59908227212
Minibatch step 9/100: mean batch inertia: 89097.66750295844, ewa inertia: 89097.66750295844
Minibatch s

C:\Users\Brech\anaconda3\envs\chidi\lib\site-packages\sklearn\cluster\_kmeans.py:1043: UserWarning: MiniBatchKMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can prevent it by setting batch_size >= 3072 or by setting the environment variable OMP_NUM_THREADS=9
  warnings.warn(


Init 1/3 with method k-means++
Inertia for init 1/3: 290094516.0
Init 2/3 with method k-means++
Inertia for init 2/3: 289787921.0
Init 3/3 with method k-means++
Inertia for init 3/3: 287686805.0
Minibatch step 1/100: mean batch inertia: 137482.52752513165
Minibatch step 2/100: mean batch inertia: 91326.69012835741, ewa inertia: 91326.69012835741
Minibatch step 3/100: mean batch inertia: 89869.42891218173, ewa inertia: 89869.42891218173
Minibatch step 4/100: mean batch inertia: 90503.86162197443, ewa inertia: 90503.86162197443
Minibatch step 5/100: mean batch inertia: 89897.60839785408, ewa inertia: 89897.60839785408
Minibatch step 6/100: mean batch inertia: 89321.58770562062, ewa inertia: 89321.58770562062
Minibatch step 7/100: mean batch inertia: 88301.14765053567, ewa inertia: 88301.14765053567
Minibatch step 8/100: mean batch inertia: 87844.30806116496, ewa inertia: 87844.30806116496
Minibatch step 9/100: mean batch inertia: 89066.652212564, ewa inertia: 89066.652212564
Minibatch st

C:\Users\Brech\anaconda3\envs\chidi\lib\site-packages\sklearn\cluster\_kmeans.py:1043: UserWarning: MiniBatchKMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can prevent it by setting batch_size >= 3072 or by setting the environment variable OMP_NUM_THREADS=9
  warnings.warn(


Init 1/3 with method k-means++
Inertia for init 1/3: 291144905.0
Init 2/3 with method k-means++
Inertia for init 2/3: 287004233.0
Init 3/3 with method k-means++
Inertia for init 3/3: 285765977.0
Minibatch step 1/100: mean batch inertia: 137145.33317376734
Minibatch step 2/100: mean batch inertia: 92854.05226556977, ewa inertia: 92854.05226556977
Minibatch step 3/100: mean batch inertia: 90264.91441180793, ewa inertia: 90264.91441180793
Minibatch step 4/100: mean batch inertia: 89801.17039382568, ewa inertia: 89801.17039382568
Minibatch step 5/100: mean batch inertia: 89472.67083116106, ewa inertia: 89472.67083116106
Minibatch step 6/100: mean batch inertia: 88689.62854156471, ewa inertia: 88689.62854156471
Minibatch step 7/100: mean batch inertia: 89366.68319189007, ewa inertia: 89366.68319189007
Minibatch step 8/100: mean batch inertia: 88684.28480200049, ewa inertia: 88684.28480200049
Minibatch step 9/100: mean batch inertia: 88312.06218168036, ewa inertia: 88312.06218168036
Minibatc

C:\Users\Brech\anaconda3\envs\chidi\lib\site-packages\sklearn\cluster\_kmeans.py:1043: UserWarning: MiniBatchKMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can prevent it by setting batch_size >= 3072 or by setting the environment variable OMP_NUM_THREADS=9
  warnings.warn(


Init 1/3 with method k-means++
Inertia for init 1/3: 284589921.0
Init 2/3 with method k-means++
Inertia for init 2/3: 286242553.0
Init 3/3 with method k-means++
Inertia for init 3/3: 292588252.0
Minibatch step 1/100: mean batch inertia: 137254.073719483
Minibatch step 2/100: mean batch inertia: 92071.92777098565, ewa inertia: 92071.92777098565
Minibatch step 3/100: mean batch inertia: 91223.90046926138, ewa inertia: 91223.90046926138
Minibatch step 4/100: mean batch inertia: 89656.02176311446, ewa inertia: 89656.02176311446
Minibatch step 5/100: mean batch inertia: 88710.90038557287, ewa inertia: 88710.90038557287
Minibatch step 6/100: mean batch inertia: 88958.45026407012, ewa inertia: 88958.45026407012
Minibatch step 7/100: mean batch inertia: 89025.89623848416, ewa inertia: 89025.89623848416
Minibatch step 8/100: mean batch inertia: 89262.28150180135, ewa inertia: 89262.28150180135
Minibatch step 9/100: mean batch inertia: 89151.67427949088, ewa inertia: 89151.67427949088
Minibatch 

C:\Users\Brech\anaconda3\envs\chidi\lib\site-packages\sklearn\cluster\_kmeans.py:1043: UserWarning: MiniBatchKMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can prevent it by setting batch_size >= 3072 or by setting the environment variable OMP_NUM_THREADS=9
  warnings.warn(


Init 1/3 with method k-means++
Inertia for init 1/3: 288385976.0
Init 2/3 with method k-means++
Inertia for init 2/3: 286518330.0
Init 3/3 with method k-means++
Inertia for init 3/3: 290627477.0
Minibatch step 1/100: mean batch inertia: 137753.1416945907
Minibatch step 2/100: mean batch inertia: 91800.80291509944, ewa inertia: 91800.80291509944
Minibatch step 3/100: mean batch inertia: 90329.39230129578, ewa inertia: 90329.39230129578
Minibatch step 4/100: mean batch inertia: 89669.9597004948, ewa inertia: 89669.9597004948
Minibatch step 5/100: mean batch inertia: 88132.3499663287, ewa inertia: 88132.3499663287
Minibatch step 6/100: mean batch inertia: 88881.51803582688, ewa inertia: 88881.51803582688
Minibatch step 7/100: mean batch inertia: 89260.48171531716, ewa inertia: 89260.48171531716
Minibatch step 8/100: mean batch inertia: 88639.23727034754, ewa inertia: 88639.23727034754
Minibatch step 9/100: mean batch inertia: 88416.41719836152, ewa inertia: 88416.41719836152
Minibatch ste

C:\Users\Brech\anaconda3\envs\chidi\lib\site-packages\sklearn\cluster\_kmeans.py:1043: UserWarning: MiniBatchKMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can prevent it by setting batch_size >= 3072 or by setting the environment variable OMP_NUM_THREADS=9
  warnings.warn(


Init 1/3 with method k-means++
Inertia for init 1/3: 289621293.0
Init 2/3 with method k-means++
Inertia for init 2/3: 281987887.0
Init 3/3 with method k-means++
Inertia for init 3/3: 290178181.0
Minibatch step 1/100: mean batch inertia: 133783.5753949258
Minibatch step 2/100: mean batch inertia: 91431.49306446059, ewa inertia: 91431.49306446059
Minibatch step 3/100: mean batch inertia: 90467.08505692621, ewa inertia: 90467.08505692621
Minibatch step 4/100: mean batch inertia: 89612.79645692492, ewa inertia: 89612.79645692492
Minibatch step 5/100: mean batch inertia: 89324.76728488338, ewa inertia: 89324.76728488338
Minibatch step 6/100: mean batch inertia: 88310.22245100123, ewa inertia: 88310.22245100123
Minibatch step 7/100: mean batch inertia: 88984.04289271917, ewa inertia: 88984.04289271917
Minibatch step 8/100: mean batch inertia: 89670.17020087203, ewa inertia: 89670.17020087203
Minibatch step 9/100: mean batch inertia: 88777.99528699744, ewa inertia: 88777.99528699744
Minibatch

C:\Users\Brech\anaconda3\envs\chidi\lib\site-packages\sklearn\cluster\_kmeans.py:1043: UserWarning: MiniBatchKMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can prevent it by setting batch_size >= 3072 or by setting the environment variable OMP_NUM_THREADS=9
  warnings.warn(


Init 1/3 with method k-means++
Inertia for init 1/3: 289247069.0
Init 2/3 with method k-means++
Inertia for init 2/3: 285371968.0
Init 3/3 with method k-means++
Inertia for init 3/3: 287155615.0
Minibatch step 1/100: mean batch inertia: 137408.50263283867
Minibatch step 2/100: mean batch inertia: 92348.12862367659, ewa inertia: 92348.12862367659
Minibatch step 3/100: mean batch inertia: 90531.11398697953, ewa inertia: 90531.11398697953
Minibatch step 4/100: mean batch inertia: 88364.61687939147, ewa inertia: 88364.61687939147
Minibatch step 5/100: mean batch inertia: 89122.29761838025, ewa inertia: 89122.29761838025
Minibatch step 6/100: mean batch inertia: 88166.23358685743, ewa inertia: 88166.23358685743
Minibatch step 7/100: mean batch inertia: 89640.68645661174, ewa inertia: 89640.68645661174
Minibatch step 8/100: mean batch inertia: 89505.90694187199, ewa inertia: 89505.90694187199
Minibatch step 9/100: mean batch inertia: 88288.56952842811, ewa inertia: 88288.56952842811
Minibatc

C:\Users\Brech\anaconda3\envs\chidi\lib\site-packages\sklearn\cluster\_kmeans.py:1043: UserWarning: MiniBatchKMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can prevent it by setting batch_size >= 3072 or by setting the environment variable OMP_NUM_THREADS=9
  warnings.warn(


Init 1/3 with method k-means++
Inertia for init 1/3: 281975340.0
Init 2/3 with method k-means++
Inertia for init 2/3: 286841639.0
Init 3/3 with method k-means++
Inertia for init 3/3: 286113833.0
Minibatch step 1/100: mean batch inertia: 136755.270464337
Minibatch step 2/100: mean batch inertia: 90833.33186858262, ewa inertia: 90833.33186858262
Minibatch step 3/100: mean batch inertia: 89216.17177501136, ewa inertia: 89216.17177501136
Minibatch step 4/100: mean batch inertia: 89244.5059233639, ewa inertia: 89244.5059233639
Minibatch step 5/100: mean batch inertia: 89803.92399729507, ewa inertia: 89803.92399729507
Minibatch step 6/100: mean batch inertia: 88963.1619192075, ewa inertia: 88963.1619192075
Minibatch step 7/100: mean batch inertia: 87839.21181041587, ewa inertia: 87839.21181041587
Minibatch step 8/100: mean batch inertia: 87675.60596921941, ewa inertia: 87675.60596921941
Minibatch step 9/100: mean batch inertia: 88676.76634729389, ewa inertia: 88676.76634729389
Minibatch step

C:\Users\Brech\anaconda3\envs\chidi\lib\site-packages\sklearn\cluster\_kmeans.py:1043: UserWarning: MiniBatchKMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can prevent it by setting batch_size >= 3072 or by setting the environment variable OMP_NUM_THREADS=9
  warnings.warn(


Init 1/3 with method k-means++
Inertia for init 1/3: 293339209.0
Init 2/3 with method k-means++
Inertia for init 2/3: 290775822.0
Init 3/3 with method k-means++
Inertia for init 3/3: 289575386.0
Minibatch step 1/100: mean batch inertia: 137700.10674964098
Minibatch step 2/100: mean batch inertia: 91410.57647416476, ewa inertia: 91410.57647416476
Minibatch step 3/100: mean batch inertia: 90699.61938657926, ewa inertia: 90699.61938657926
Minibatch step 4/100: mean batch inertia: 89582.96619829396, ewa inertia: 89582.96619829396
Minibatch step 5/100: mean batch inertia: 88412.34884691285, ewa inertia: 88412.34884691285
Minibatch step 6/100: mean batch inertia: 88587.67087536302, ewa inertia: 88587.67087536302
Minibatch step 7/100: mean batch inertia: 88270.60867650213, ewa inertia: 88270.60867650213
Minibatch step 8/100: mean batch inertia: 87996.73533845962, ewa inertia: 87996.73533845962
Minibatch step 9/100: mean batch inertia: 87650.5351780118, ewa inertia: 87650.5351780118
Minibatch 

C:\Users\Brech\anaconda3\envs\chidi\lib\site-packages\sklearn\cluster\_kmeans.py:1043: UserWarning: MiniBatchKMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can prevent it by setting batch_size >= 3072 or by setting the environment variable OMP_NUM_THREADS=9
  warnings.warn(


Init 1/3 with method k-means++
Inertia for init 1/3: 282406026.0
Init 2/3 with method k-means++
Inertia for init 2/3: 290761718.0
Init 3/3 with method k-means++
Inertia for init 3/3: 284958069.0
Minibatch step 1/100: mean batch inertia: 136930.95117280996
Minibatch step 2/100: mean batch inertia: 91473.89275044097, ewa inertia: 91473.89275044097
Minibatch step 3/100: mean batch inertia: 90691.61011050109, ewa inertia: 90691.61011050109
Minibatch step 4/100: mean batch inertia: 90133.81138413447, ewa inertia: 90133.81138413447
Minibatch step 5/100: mean batch inertia: 88972.99181948758, ewa inertia: 88972.99181948758
Minibatch step 6/100: mean batch inertia: 89848.2657080337, ewa inertia: 89848.2657080337
Minibatch step 7/100: mean batch inertia: 89505.45297038827, ewa inertia: 89505.45297038827
Minibatch step 8/100: mean batch inertia: 88727.91588861865, ewa inertia: 88727.91588861865
Minibatch step 9/100: mean batch inertia: 87975.63536447428, ewa inertia: 87975.63536447428
Minibatch 

C:\Users\Brech\anaconda3\envs\chidi\lib\site-packages\sklearn\cluster\_kmeans.py:1043: UserWarning: MiniBatchKMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can prevent it by setting batch_size >= 3072 or by setting the environment variable OMP_NUM_THREADS=9
  warnings.warn(


Init 1/3 with method k-means++
Inertia for init 1/3: 287739630.0
Init 2/3 with method k-means++
Inertia for init 2/3: 296673843.0
Init 3/3 with method k-means++
Inertia for init 3/3: 296168978.0
Minibatch step 1/100: mean batch inertia: 136458.99138343704
Minibatch step 2/100: mean batch inertia: 91619.20471484259, ewa inertia: 91619.20471484259
Minibatch step 3/100: mean batch inertia: 90027.76808462452, ewa inertia: 90027.76808462452
Minibatch step 4/100: mean batch inertia: 89202.4249494724, ewa inertia: 89202.4249494724
Minibatch step 5/100: mean batch inertia: 88894.61154610128, ewa inertia: 88894.61154610128
Minibatch step 6/100: mean batch inertia: 88477.8228144067, ewa inertia: 88477.8228144067
Minibatch step 7/100: mean batch inertia: 88182.84624156344, ewa inertia: 88182.84624156344
Minibatch step 8/100: mean batch inertia: 89175.14362486781, ewa inertia: 89175.14362486781
Minibatch step 9/100: mean batch inertia: 89177.2418336143, ewa inertia: 89177.2418336143
Minibatch step

C:\Users\Brech\anaconda3\envs\chidi\lib\site-packages\sklearn\cluster\_kmeans.py:1043: UserWarning: MiniBatchKMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can prevent it by setting batch_size >= 3072 or by setting the environment variable OMP_NUM_THREADS=9
  warnings.warn(


Init 1/3 with method k-means++
Inertia for init 1/3: 282988516.0
Init 2/3 with method k-means++
Inertia for init 2/3: 287417615.0
Init 3/3 with method k-means++
Inertia for init 3/3: 284913004.0
Minibatch step 1/100: mean batch inertia: 134245.3106749641
Minibatch step 2/100: mean batch inertia: 91907.02823824348, ewa inertia: 91907.02823824348
Minibatch step 3/100: mean batch inertia: 88709.35694059469, ewa inertia: 88709.35694059469
Minibatch step 4/100: mean batch inertia: 89146.34479643124, ewa inertia: 89146.34479643124
Minibatch step 5/100: mean batch inertia: 88774.75453053691, ewa inertia: 88774.75453053691
Minibatch step 6/100: mean batch inertia: 88059.42524072337, ewa inertia: 88059.42524072337
Minibatch step 7/100: mean batch inertia: 88380.91708561094, ewa inertia: 88380.91708561094
Minibatch step 8/100: mean batch inertia: 88260.17257036731, ewa inertia: 88260.17257036731
Minibatch step 9/100: mean batch inertia: 87803.2997919694, ewa inertia: 87803.2997919694
Minibatch s

C:\Users\Brech\anaconda3\envs\chidi\lib\site-packages\sklearn\cluster\_kmeans.py:1043: UserWarning: MiniBatchKMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can prevent it by setting batch_size >= 3072 or by setting the environment variable OMP_NUM_THREADS=9
  warnings.warn(


Init 1/3 with method k-means++
Inertia for init 1/3: 287257084.0
Init 2/3 with method k-means++
Inertia for init 2/3: 286637427.0
Init 3/3 with method k-means++
Inertia for init 3/3: 289857488.0
Minibatch step 1/100: mean batch inertia: 136610.91670655817
Minibatch step 2/100: mean batch inertia: 91896.17623180024, ewa inertia: 91896.17623180024
Minibatch step 3/100: mean batch inertia: 89666.60589201622, ewa inertia: 89666.60589201622
Minibatch step 4/100: mean batch inertia: 89729.9404085306, ewa inertia: 89729.9404085306
Minibatch step 5/100: mean batch inertia: 90166.62527477316, ewa inertia: 90166.62527477316
Minibatch step 6/100: mean batch inertia: 89067.96702832272, ewa inertia: 89067.96702832272
Minibatch step 7/100: mean batch inertia: 89105.62514552905, ewa inertia: 89105.62514552905
Minibatch step 8/100: mean batch inertia: 88477.58979928405, ewa inertia: 88477.58979928405
Minibatch step 9/100: mean batch inertia: 88481.0159080736, ewa inertia: 88481.0159080736
Minibatch st

C:\Users\Brech\anaconda3\envs\chidi\lib\site-packages\sklearn\cluster\_kmeans.py:1043: UserWarning: MiniBatchKMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can prevent it by setting batch_size >= 3072 or by setting the environment variable OMP_NUM_THREADS=9
  warnings.warn(


Init 1/3 with method k-means++
Inertia for init 1/3: 285100464.0
Init 2/3 with method k-means++
Inertia for init 2/3: 284866640.0
Init 3/3 with method k-means++
Inertia for init 3/3: 284635710.0
Minibatch step 1/100: mean batch inertia: 136803.89564384872
Minibatch step 2/100: mean batch inertia: 90813.31151160544, ewa inertia: 90813.31151160544
Minibatch step 3/100: mean batch inertia: 89113.10937566782, ewa inertia: 89113.10937566782
Minibatch step 4/100: mean batch inertia: 89146.82790084305, ewa inertia: 89146.82790084305
Minibatch step 5/100: mean batch inertia: 88020.66595731268, ewa inertia: 88020.66595731268
Minibatch step 6/100: mean batch inertia: 89323.89410209461, ewa inertia: 89323.89410209461
Minibatch step 7/100: mean batch inertia: 88230.32452080594, ewa inertia: 88230.32452080594
Minibatch step 8/100: mean batch inertia: 87395.0771964278, ewa inertia: 87395.0771964278
Minibatch step 9/100: mean batch inertia: 88367.03580843387, ewa inertia: 88367.03580843387
Minibatch 

C:\Users\Brech\anaconda3\envs\chidi\lib\site-packages\sklearn\cluster\_kmeans.py:1043: UserWarning: MiniBatchKMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can prevent it by setting batch_size >= 3072 or by setting the environment variable OMP_NUM_THREADS=9
  warnings.warn(


Init 1/3 with method k-means++
Inertia for init 1/3: 289158087.0
Init 2/3 with method k-means++
Inertia for init 2/3: 289181500.0
Init 3/3 with method k-means++
Inertia for init 3/3: 291353513.0
Minibatch step 1/100: mean batch inertia: 136789.88989947343
Minibatch step 2/100: mean batch inertia: 90941.72650383526, ewa inertia: 90941.72650383526
Minibatch step 3/100: mean batch inertia: 89491.92347831304, ewa inertia: 89491.92347831304
Minibatch step 4/100: mean batch inertia: 88987.07712676015, ewa inertia: 88987.07712676015
Minibatch step 5/100: mean batch inertia: 88601.1051995569, ewa inertia: 88601.1051995569
Minibatch step 6/100: mean batch inertia: 88405.27888766254, ewa inertia: 88405.27888766254
Minibatch step 7/100: mean batch inertia: 89430.77425668965, ewa inertia: 89430.77425668965
Minibatch step 8/100: mean batch inertia: 88392.42830942017, ewa inertia: 88392.42830942017
Minibatch step 9/100: mean batch inertia: 88013.03552075094, ewa inertia: 88013.03552075094
Minibatch 

C:\Users\Brech\anaconda3\envs\chidi\lib\site-packages\sklearn\cluster\_kmeans.py:1043: UserWarning: MiniBatchKMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can prevent it by setting batch_size >= 3072 or by setting the environment variable OMP_NUM_THREADS=9
  warnings.warn(


Init 1/3 with method k-means++
Inertia for init 1/3: 286383809.0
Init 2/3 with method k-means++
Inertia for init 2/3: 293471868.0
Init 3/3 with method k-means++
Inertia for init 3/3: 294177053.0
Minibatch step 1/100: mean batch inertia: 136914.7826711345
Minibatch step 2/100: mean batch inertia: 92167.0775754708, ewa inertia: 92167.0775754708
Minibatch step 3/100: mean batch inertia: 90852.5074840087, ewa inertia: 90852.5074840087
Minibatch step 4/100: mean batch inertia: 90278.97514795497, ewa inertia: 90278.97514795497
Minibatch step 5/100: mean batch inertia: 88876.55290098792, ewa inertia: 88876.55290098792
Minibatch step 6/100: mean batch inertia: 89215.23083447211, ewa inertia: 89215.23083447211
Minibatch step 7/100: mean batch inertia: 88858.72018358244, ewa inertia: 88858.72018358244
Minibatch step 8/100: mean batch inertia: 88152.18112559716, ewa inertia: 88152.18112559716
Minibatch step 9/100: mean batch inertia: 89145.69008449794, ewa inertia: 89145.69008449794
Minibatch ste

C:\Users\Brech\anaconda3\envs\chidi\lib\site-packages\sklearn\cluster\_kmeans.py:1043: UserWarning: MiniBatchKMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can prevent it by setting batch_size >= 3072 or by setting the environment variable OMP_NUM_THREADS=9
  warnings.warn(


Init 1/3 with method k-means++
Inertia for init 1/3: 288499697.0
Init 2/3 with method k-means++
Inertia for init 2/3: 289726128.0
Init 3/3 with method k-means++
Inertia for init 3/3: 286924913.0
Minibatch step 1/100: mean batch inertia: 138114.91957874582
Minibatch step 2/100: mean batch inertia: 91175.07466847963, ewa inertia: 91175.07466847963
Minibatch step 3/100: mean batch inertia: 90951.94485960217, ewa inertia: 90951.94485960217
Minibatch step 4/100: mean batch inertia: 89406.3435083594, ewa inertia: 89406.3435083594
Minibatch step 5/100: mean batch inertia: 89631.93741372136, ewa inertia: 89631.93741372136
Minibatch step 6/100: mean batch inertia: 89097.00505471375, ewa inertia: 89097.00505471375
Minibatch step 7/100: mean batch inertia: 88991.28409301476, ewa inertia: 88991.28409301476
Minibatch step 8/100: mean batch inertia: 88676.46038501887, ewa inertia: 88676.46038501887
Minibatch step 9/100: mean batch inertia: 88042.79217333038, ewa inertia: 88042.79217333038
Minibatch 

C:\Users\Brech\anaconda3\envs\chidi\lib\site-packages\sklearn\cluster\_kmeans.py:1043: UserWarning: MiniBatchKMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can prevent it by setting batch_size >= 3072 or by setting the environment variable OMP_NUM_THREADS=9
  warnings.warn(


Init 1/3 with method k-means++
Inertia for init 1/3: 283409899.0
Init 2/3 with method k-means++
Inertia for init 2/3: 280413059.0
Init 3/3 with method k-means++
Inertia for init 3/3: 288320691.0
Minibatch step 1/100: mean batch inertia: 135238.18621349928
Minibatch step 2/100: mean batch inertia: 90817.53612718623, ewa inertia: 90817.53612718623
Minibatch step 3/100: mean batch inertia: 89610.35926044318, ewa inertia: 89610.35926044318
Minibatch step 4/100: mean batch inertia: 88619.61295675772, ewa inertia: 88619.61295675772
Minibatch step 5/100: mean batch inertia: 89594.07093947913, ewa inertia: 89594.07093947913
Minibatch step 6/100: mean batch inertia: 89357.64485951755, ewa inertia: 89357.64485951755
Minibatch step 7/100: mean batch inertia: 88973.13461037338, ewa inertia: 88973.13461037338
Minibatch step 8/100: mean batch inertia: 88387.2672075863, ewa inertia: 88387.2672075863
Minibatch step 9/100: mean batch inertia: 87629.77254867861, ewa inertia: 87629.77254867861
Minibatch 

C:\Users\Brech\anaconda3\envs\chidi\lib\site-packages\sklearn\cluster\_kmeans.py:1043: UserWarning: MiniBatchKMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can prevent it by setting batch_size >= 3072 or by setting the environment variable OMP_NUM_THREADS=9
  warnings.warn(


Init 1/3 with method k-means++
Inertia for init 1/3: 286690477.0
Init 2/3 with method k-means++
Inertia for init 2/3: 289248907.0
Init 3/3 with method k-means++
Inertia for init 3/3: 288268772.0
Minibatch step 1/100: mean batch inertia: 138015.02297750118
Minibatch step 2/100: mean batch inertia: 90082.3006906558, ewa inertia: 90082.3006906558
Minibatch step 3/100: mean batch inertia: 89176.82043871205, ewa inertia: 89176.82043871205
Minibatch step 4/100: mean batch inertia: 89086.48984969768, ewa inertia: 89086.48984969768
Minibatch step 5/100: mean batch inertia: 89276.68248956223, ewa inertia: 89276.68248956223
Minibatch step 6/100: mean batch inertia: 88916.84576196624, ewa inertia: 88916.84576196624
Minibatch step 7/100: mean batch inertia: 88941.42689786266, ewa inertia: 88941.42689786266
Minibatch step 8/100: mean batch inertia: 88133.9681780059, ewa inertia: 88133.9681780059
Minibatch step 9/100: mean batch inertia: 88647.68382888062, ewa inertia: 88647.68382888062
Minibatch st

C:\Users\Brech\anaconda3\envs\chidi\lib\site-packages\sklearn\cluster\_kmeans.py:1043: UserWarning: MiniBatchKMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can prevent it by setting batch_size >= 3072 or by setting the environment variable OMP_NUM_THREADS=9
  warnings.warn(


Init 1/3 with method k-means++
Inertia for init 1/3: 288146023.0
Init 2/3 with method k-means++
Inertia for init 2/3: 284646719.0
Init 3/3 with method k-means++
Inertia for init 3/3: 285987467.0
Minibatch step 1/100: mean batch inertia: 138620.58161799904
Minibatch step 2/100: mean batch inertia: 92818.14170820035, ewa inertia: 92818.14170820035
Minibatch step 3/100: mean batch inertia: 90516.18075179517, ewa inertia: 90516.18075179517
Minibatch step 4/100: mean batch inertia: 89250.29809523174, ewa inertia: 89250.29809523174
Minibatch step 5/100: mean batch inertia: 89209.54520746222, ewa inertia: 89209.54520746222
Minibatch step 6/100: mean batch inertia: 89764.21065334805, ewa inertia: 89764.21065334805
Minibatch step 7/100: mean batch inertia: 89065.89907339051, ewa inertia: 89065.89907339051
Minibatch step 8/100: mean batch inertia: 89584.84601223565, ewa inertia: 89584.84601223565
Minibatch step 9/100: mean batch inertia: 88238.26884399573, ewa inertia: 88238.26884399573
Minibatc

C:\Users\Brech\anaconda3\envs\chidi\lib\site-packages\sklearn\cluster\_kmeans.py:1043: UserWarning: MiniBatchKMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can prevent it by setting batch_size >= 3072 or by setting the environment variable OMP_NUM_THREADS=9
  warnings.warn(


Init 1/3 with method k-means++
Inertia for init 1/3: 287408309.0
Init 2/3 with method k-means++
Inertia for init 2/3: 291016485.0
Init 3/3 with method k-means++
Inertia for init 3/3: 291201216.0
Minibatch step 1/100: mean batch inertia: 137756.59167065582
Minibatch step 2/100: mean batch inertia: 91761.7539787352, ewa inertia: 91761.7539787352
Minibatch step 3/100: mean batch inertia: 90308.97438383155, ewa inertia: 90308.97438383155
Minibatch step 4/100: mean batch inertia: 89848.16562965586, ewa inertia: 89848.16562965586
Minibatch step 5/100: mean batch inertia: 88794.6328188368, ewa inertia: 88794.6328188368
Minibatch step 6/100: mean batch inertia: 87842.09807202985, ewa inertia: 87842.09807202985
Minibatch step 7/100: mean batch inertia: 88621.97885289036, ewa inertia: 88621.97885289036
Minibatch step 8/100: mean batch inertia: 88340.53875340507, ewa inertia: 88340.53875340507
Minibatch step 9/100: mean batch inertia: 88159.31007367175, ewa inertia: 88159.31007367175
Minibatch st

C:\Users\Brech\anaconda3\envs\chidi\lib\site-packages\sklearn\cluster\_kmeans.py:1043: UserWarning: MiniBatchKMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can prevent it by setting batch_size >= 3072 or by setting the environment variable OMP_NUM_THREADS=9
  warnings.warn(


Init 1/3 with method k-means++
Inertia for init 1/3: 287118853.0
Init 2/3 with method k-means++
Inertia for init 2/3: 285677694.0
Init 3/3 with method k-means++
Inertia for init 3/3: 284346677.0
Minibatch step 1/100: mean batch inertia: 137094.51268549546
Minibatch step 2/100: mean batch inertia: 92484.09384095333, ewa inertia: 92484.09384095333
Minibatch step 3/100: mean batch inertia: 89993.27882879351, ewa inertia: 89993.27882879351
Minibatch step 4/100: mean batch inertia: 90350.56362490254, ewa inertia: 90350.56362490254
Minibatch step 5/100: mean batch inertia: 89094.27322716272, ewa inertia: 89094.27322716272
Minibatch step 6/100: mean batch inertia: 89147.34223180357, ewa inertia: 89147.34223180357
Minibatch step 7/100: mean batch inertia: 88248.05456682312, ewa inertia: 88248.05456682312
Minibatch step 8/100: mean batch inertia: 88844.85929964433, ewa inertia: 88844.85929964433
Minibatch step 9/100: mean batch inertia: 88891.3041897627, ewa inertia: 88891.3041897627
Minibatch 

C:\Users\Brech\anaconda3\envs\chidi\lib\site-packages\sklearn\cluster\_kmeans.py:1043: UserWarning: MiniBatchKMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can prevent it by setting batch_size >= 3072 or by setting the environment variable OMP_NUM_THREADS=9
  warnings.warn(


Init 1/3 with method k-means++
Inertia for init 1/3: 289322914.0
Init 2/3 with method k-means++
Inertia for init 2/3: 285691625.0
Init 3/3 with method k-means++
Inertia for init 3/3: 284179442.0
Minibatch step 1/100: mean batch inertia: 137855.5974150311
Minibatch step 2/100: mean batch inertia: 91279.20380529399, ewa inertia: 91279.20380529399
Minibatch step 3/100: mean batch inertia: 89189.75965575674, ewa inertia: 89189.75965575674
Minibatch step 4/100: mean batch inertia: 88229.19554867336, ewa inertia: 88229.19554867336
Minibatch step 5/100: mean batch inertia: 88762.99933985513, ewa inertia: 88762.99933985513
Minibatch step 6/100: mean batch inertia: 88152.87063121896, ewa inertia: 88152.87063121896
Minibatch step 7/100: mean batch inertia: 88467.31428986146, ewa inertia: 88467.31428986146
Minibatch step 8/100: mean batch inertia: 87733.98566359859, ewa inertia: 87733.98566359859
Minibatch step 9/100: mean batch inertia: 88195.00441835457, ewa inertia: 88195.00441835457
Minibatch

C:\Users\Brech\anaconda3\envs\chidi\lib\site-packages\sklearn\cluster\_kmeans.py:1043: UserWarning: MiniBatchKMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can prevent it by setting batch_size >= 3072 or by setting the environment variable OMP_NUM_THREADS=9
  warnings.warn(


Init 1/3 with method k-means++
Inertia for init 1/3: 291688479.0
Init 2/3 with method k-means++
Inertia for init 2/3: 290627266.0
Init 3/3 with method k-means++
Inertia for init 3/3: 291822999.0
Minibatch step 1/100: mean batch inertia: 138150.94638583055
Minibatch step 2/100: mean batch inertia: 91629.16083496276, ewa inertia: 91629.16083496276
Minibatch step 3/100: mean batch inertia: 89505.25590660455, ewa inertia: 89505.25590660455
Minibatch step 4/100: mean batch inertia: 90456.38558506702, ewa inertia: 90456.38558506702
Minibatch step 5/100: mean batch inertia: 89061.17967676291, ewa inertia: 89061.17967676291
Minibatch step 6/100: mean batch inertia: 89233.78976081348, ewa inertia: 89233.78976081348
Minibatch step 7/100: mean batch inertia: 87793.93483970269, ewa inertia: 87793.93483970269
Minibatch step 8/100: mean batch inertia: 88442.19761749664, ewa inertia: 88442.19761749664
Minibatch step 9/100: mean batch inertia: 88947.54052612389, ewa inertia: 88947.54052612389
Minibatc

C:\Users\Brech\anaconda3\envs\chidi\lib\site-packages\sklearn\cluster\_kmeans.py:1043: UserWarning: MiniBatchKMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can prevent it by setting batch_size >= 3072 or by setting the environment variable OMP_NUM_THREADS=9
  warnings.warn(


Init 1/3 with method k-means++
Inertia for init 1/3: 286698303.0
Init 2/3 with method k-means++
Inertia for init 2/3: 287480839.0
Init 3/3 with method k-means++
Inertia for init 3/3: 283467676.0
Minibatch step 1/100: mean batch inertia: 136485.1081857348
Minibatch step 2/100: mean batch inertia: 91537.4472274848, ewa inertia: 91537.4472274848
Minibatch step 3/100: mean batch inertia: 89383.23442722365, ewa inertia: 89383.23442722365
Minibatch step 4/100: mean batch inertia: 90811.76052067034, ewa inertia: 90811.76052067034
Minibatch step 5/100: mean batch inertia: 88787.16772017026, ewa inertia: 88787.16772017026
Minibatch step 6/100: mean batch inertia: 88533.65242372165, ewa inertia: 88533.65242372165
Minibatch step 7/100: mean batch inertia: 88291.08209482745, ewa inertia: 88291.08209482745
Minibatch step 8/100: mean batch inertia: 88205.53150758558, ewa inertia: 88205.53150758558
Minibatch step 9/100: mean batch inertia: 88644.71669571946, ewa inertia: 88644.71669571946
Minibatch s

C:\Users\Brech\anaconda3\envs\chidi\lib\site-packages\sklearn\cluster\_kmeans.py:1043: UserWarning: MiniBatchKMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can prevent it by setting batch_size >= 3072 or by setting the environment variable OMP_NUM_THREADS=9
  warnings.warn(


Init 1/3 with method k-means++
Inertia for init 1/3: 278336484.0
Init 2/3 with method k-means++
Inertia for init 2/3: 286726823.0
Init 3/3 with method k-means++
Inertia for init 3/3: 283119455.0
Minibatch step 1/100: mean batch inertia: 135005.52225945427
Minibatch step 2/100: mean batch inertia: 92010.77375604887, ewa inertia: 92010.77375604887
Minibatch step 3/100: mean batch inertia: 89839.7024733946, ewa inertia: 89839.7024733946
Minibatch step 4/100: mean batch inertia: 90070.90456052689, ewa inertia: 90070.90456052689
Minibatch step 5/100: mean batch inertia: 88858.76016213695, ewa inertia: 88858.76016213695
Minibatch step 6/100: mean batch inertia: 88275.71434257206, ewa inertia: 88275.71434257206
Minibatch step 7/100: mean batch inertia: 88123.61614127869, ewa inertia: 88123.61614127869
Minibatch step 8/100: mean batch inertia: 88297.26737945208, ewa inertia: 88297.26737945208
Minibatch step 9/100: mean batch inertia: 87856.4741895845, ewa inertia: 87856.4741895845
Minibatch st

C:\Users\Brech\anaconda3\envs\chidi\lib\site-packages\sklearn\cluster\_kmeans.py:1043: UserWarning: MiniBatchKMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can prevent it by setting batch_size >= 3072 or by setting the environment variable OMP_NUM_THREADS=9
  warnings.warn(


Init 1/3 with method k-means++
Inertia for init 1/3: 292422148.0
Init 2/3 with method k-means++
Inertia for init 2/3: 287842627.0
Init 3/3 with method k-means++
Inertia for init 3/3: 288526819.0
Minibatch step 1/100: mean batch inertia: 137635.02728578268
Minibatch step 2/100: mean batch inertia: 91830.46258366627, ewa inertia: 91830.46258366627
Minibatch step 3/100: mean batch inertia: 90301.46945235463, ewa inertia: 90301.46945235463
Minibatch step 4/100: mean batch inertia: 89703.21340342192, ewa inertia: 89703.21340342192
Minibatch step 5/100: mean batch inertia: 88366.88922512141, ewa inertia: 88366.88922512141
Minibatch step 6/100: mean batch inertia: 89110.93009686597, ewa inertia: 89110.93009686597
Minibatch step 7/100: mean batch inertia: 89270.84523664304, ewa inertia: 89270.84523664304
Minibatch step 8/100: mean batch inertia: 89169.48989542128, ewa inertia: 89169.48989542128
Minibatch step 9/100: mean batch inertia: 89355.55740673257, ewa inertia: 89355.55740673257
Minibatc

C:\Users\Brech\anaconda3\envs\chidi\lib\site-packages\sklearn\cluster\_kmeans.py:1043: UserWarning: MiniBatchKMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can prevent it by setting batch_size >= 3072 or by setting the environment variable OMP_NUM_THREADS=9
  warnings.warn(


Init 1/3 with method k-means++
Inertia for init 1/3: 287525344.0
Init 2/3 with method k-means++
Inertia for init 2/3: 290303659.0
Init 3/3 with method k-means++
Inertia for init 3/3: 293876863.0
Minibatch step 1/100: mean batch inertia: 139951.96984202968
Minibatch step 2/100: mean batch inertia: 91177.36686387492, ewa inertia: 91177.36686387492
Minibatch step 3/100: mean batch inertia: 90152.3196335281, ewa inertia: 90152.3196335281
Minibatch step 4/100: mean batch inertia: 89662.94737322965, ewa inertia: 89662.94737322965
Minibatch step 5/100: mean batch inertia: 88821.67067166977, ewa inertia: 88821.67067166977
Minibatch step 6/100: mean batch inertia: 89198.42671279307, ewa inertia: 89198.42671279307
Minibatch step 7/100: mean batch inertia: 89146.53144749237, ewa inertia: 89146.53144749237
Minibatch step 8/100: mean batch inertia: 88541.05104809697, ewa inertia: 88541.05104809697
Minibatch step 9/100: mean batch inertia: 88995.67398952374, ewa inertia: 88995.67398952374
Minibatch 

C:\Users\Brech\anaconda3\envs\chidi\lib\site-packages\sklearn\cluster\_kmeans.py:1043: UserWarning: MiniBatchKMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can prevent it by setting batch_size >= 3072 or by setting the environment variable OMP_NUM_THREADS=9
  warnings.warn(


Init 1/3 with method k-means++
Inertia for init 1/3: 283292983.0
Init 2/3 with method k-means++
Inertia for init 2/3: 281745692.0
Init 3/3 with method k-means++
Inertia for init 3/3: 280683542.0
Minibatch step 1/100: mean batch inertia: 134627.35088559118
Minibatch step 2/100: mean batch inertia: 90150.48931856267, ewa inertia: 90150.48931856267
Minibatch step 3/100: mean batch inertia: 89713.00661655894, ewa inertia: 89713.00661655894
Minibatch step 4/100: mean batch inertia: 89818.03645767926, ewa inertia: 89818.03645767926
Minibatch step 5/100: mean batch inertia: 88857.21057406692, ewa inertia: 88857.21057406692
Minibatch step 6/100: mean batch inertia: 89003.05662926407, ewa inertia: 89003.05662926407
Minibatch step 7/100: mean batch inertia: 88947.49603219367, ewa inertia: 88947.49603219367
Minibatch step 8/100: mean batch inertia: 89192.15669374546, ewa inertia: 89192.15669374546
Minibatch step 9/100: mean batch inertia: 89008.85506382522, ewa inertia: 89008.85506382522
Minibatc

C:\Users\Brech\anaconda3\envs\chidi\lib\site-packages\sklearn\cluster\_kmeans.py:1043: UserWarning: MiniBatchKMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can prevent it by setting batch_size >= 3072 or by setting the environment variable OMP_NUM_THREADS=9
  warnings.warn(


Init 1/3 with method k-means++
Inertia for init 1/3: 291275214.0
Init 2/3 with method k-means++
Inertia for init 2/3: 292911649.0
Init 3/3 with method k-means++
Inertia for init 3/3: 285626725.0
Minibatch step 1/100: mean batch inertia: 136216.12637625658
Minibatch step 2/100: mean batch inertia: 90704.28713737996, ewa inertia: 90704.28713737996
Minibatch step 3/100: mean batch inertia: 89694.4659450143, ewa inertia: 89694.4659450143
Minibatch step 4/100: mean batch inertia: 89198.7006470788, ewa inertia: 89198.7006470788
Minibatch step 5/100: mean batch inertia: 89151.43091421724, ewa inertia: 89151.43091421724
Minibatch step 6/100: mean batch inertia: 88890.56189006848, ewa inertia: 88890.56189006848
Minibatch step 7/100: mean batch inertia: 88761.16200805843, ewa inertia: 88761.16200805843
Minibatch step 8/100: mean batch inertia: 88037.6277930396, ewa inertia: 88037.6277930396
Minibatch step 9/100: mean batch inertia: 87715.3538137042, ewa inertia: 87715.3538137042
Minibatch step 1

C:\Users\Brech\anaconda3\envs\chidi\lib\site-packages\sklearn\cluster\_kmeans.py:1043: UserWarning: MiniBatchKMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can prevent it by setting batch_size >= 3072 or by setting the environment variable OMP_NUM_THREADS=9
  warnings.warn(


Init 1/3 with method k-means++
Inertia for init 1/3: 291137080.0
Init 2/3 with method k-means++
Inertia for init 2/3: 288345983.0
Init 3/3 with method k-means++
Inertia for init 3/3: 289987709.0
Minibatch step 1/100: mean batch inertia: 139069.26089037818
Minibatch step 2/100: mean batch inertia: 92409.98060400499, ewa inertia: 92409.98060400499
Minibatch step 3/100: mean batch inertia: 90660.73439915954, ewa inertia: 90660.73439915954
Minibatch step 4/100: mean batch inertia: 90692.0809819683, ewa inertia: 90692.0809819683
Minibatch step 5/100: mean batch inertia: 89413.59920489811, ewa inertia: 89413.59920489811
Minibatch step 6/100: mean batch inertia: 89631.91695428951, ewa inertia: 89631.91695428951
Minibatch step 7/100: mean batch inertia: 89331.38052029753, ewa inertia: 89331.38052029753
Minibatch step 8/100: mean batch inertia: 89328.93807228972, ewa inertia: 89328.93807228972
Minibatch step 9/100: mean batch inertia: 88612.09314254251, ewa inertia: 88612.09314254251
Minibatch 

C:\Users\Brech\anaconda3\envs\chidi\lib\site-packages\sklearn\cluster\_kmeans.py:1043: UserWarning: MiniBatchKMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can prevent it by setting batch_size >= 3072 or by setting the environment variable OMP_NUM_THREADS=9
  warnings.warn(


Init 1/3 with method k-means++
Inertia for init 1/3: 286509410.0
Init 2/3 with method k-means++
Inertia for init 2/3: 288475571.0
Init 3/3 with method k-means++
Inertia for init 3/3: 283569351.0
Minibatch step 1/100: mean batch inertia: 134925.03015797032
Minibatch step 2/100: mean batch inertia: 91764.39459862442, ewa inertia: 91764.39459862442
Minibatch step 3/100: mean batch inertia: 90272.07487947485, ewa inertia: 90272.07487947485
Minibatch step 4/100: mean batch inertia: 89953.9849998422, ewa inertia: 89953.9849998422
Minibatch step 5/100: mean batch inertia: 88765.09504540403, ewa inertia: 88765.09504540403
Minibatch step 6/100: mean batch inertia: 88725.3717342213, ewa inertia: 88725.3717342213
Minibatch step 7/100: mean batch inertia: 88947.31192176358, ewa inertia: 88947.31192176358
Minibatch step 8/100: mean batch inertia: 88981.41406691824, ewa inertia: 88981.41406691824
Minibatch step 9/100: mean batch inertia: 89053.14965927752, ewa inertia: 89053.14965927752
Minibatch st

C:\Users\Brech\anaconda3\envs\chidi\lib\site-packages\sklearn\cluster\_kmeans.py:1043: UserWarning: MiniBatchKMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can prevent it by setting batch_size >= 3072 or by setting the environment variable OMP_NUM_THREADS=9
  warnings.warn(


Init 1/3 with method k-means++
Inertia for init 1/3: 290616692.0
Init 2/3 with method k-means++
Inertia for init 2/3: 288773847.0
Init 3/3 with method k-means++
Inertia for init 3/3: 287596693.0
Minibatch step 1/100: mean batch inertia: 136140.17472474868
Minibatch step 2/100: mean batch inertia: 91714.36000406521, ewa inertia: 91714.36000406521
Minibatch step 3/100: mean batch inertia: 89997.2407909244, ewa inertia: 89997.2407909244
Minibatch step 4/100: mean batch inertia: 89235.17598386541, ewa inertia: 89235.17598386541
Minibatch step 5/100: mean batch inertia: 89147.34690594998, ewa inertia: 89147.34690594998
Minibatch step 6/100: mean batch inertia: 88282.03158105268, ewa inertia: 88282.03158105268
Minibatch step 7/100: mean batch inertia: 88976.28888119991, ewa inertia: 88976.28888119991
Minibatch step 8/100: mean batch inertia: 88427.82042322567, ewa inertia: 88427.82042322567
Minibatch step 9/100: mean batch inertia: 89358.81578394382, ewa inertia: 89358.81578394382
Minibatch 

C:\Users\Brech\anaconda3\envs\chidi\lib\site-packages\sklearn\cluster\_kmeans.py:1043: UserWarning: MiniBatchKMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can prevent it by setting batch_size >= 3072 or by setting the environment variable OMP_NUM_THREADS=9
  warnings.warn(


Init 1/3 with method k-means++
Inertia for init 1/3: 282792866.0
Init 2/3 with method k-means++
Inertia for init 2/3: 287486278.0
Init 3/3 with method k-means++
Inertia for init 3/3: 284434342.0
Minibatch step 1/100: mean batch inertia: 137759.70703685973
Minibatch step 2/100: mean batch inertia: 92086.44563964858, ewa inertia: 92086.44563964858
Minibatch step 3/100: mean batch inertia: 90787.68523745016, ewa inertia: 90787.68523745016
Minibatch step 4/100: mean batch inertia: 90665.78133357449, ewa inertia: 90665.78133357449
Minibatch step 5/100: mean batch inertia: 90461.86626636017, ewa inertia: 90461.86626636017
Minibatch step 6/100: mean batch inertia: 89476.91834111688, ewa inertia: 89476.91834111688
Minibatch step 7/100: mean batch inertia: 89345.86376380544, ewa inertia: 89345.86376380544
Minibatch step 8/100: mean batch inertia: 90375.22876802956, ewa inertia: 90375.22876802956
Minibatch step 9/100: mean batch inertia: 88969.12604590341, ewa inertia: 88969.12604590341
Minibatc

C:\Users\Brech\anaconda3\envs\chidi\lib\site-packages\sklearn\cluster\_kmeans.py:1043: UserWarning: MiniBatchKMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can prevent it by setting batch_size >= 3072 or by setting the environment variable OMP_NUM_THREADS=9
  warnings.warn(


Init 1/3 with method k-means++
Inertia for init 1/3: 288219207.0
Init 2/3 with method k-means++
Inertia for init 2/3: 285348648.0
Init 3/3 with method k-means++
Inertia for init 3/3: 281733061.0
Minibatch step 1/100: mean batch inertia: 135075.9597893729
Minibatch step 2/100: mean batch inertia: 90593.6764677315, ewa inertia: 90593.6764677315
Minibatch step 3/100: mean batch inertia: 89505.44898512168, ewa inertia: 89505.44898512168
Minibatch step 4/100: mean batch inertia: 89326.09448800544, ewa inertia: 89326.09448800544
Minibatch step 5/100: mean batch inertia: 89250.24069180964, ewa inertia: 89250.24069180964
Minibatch step 6/100: mean batch inertia: 89128.81537410493, ewa inertia: 89128.81537410493
Minibatch step 7/100: mean batch inertia: 89572.28783428739, ewa inertia: 89572.28783428739
Minibatch step 8/100: mean batch inertia: 88119.87844016938, ewa inertia: 88119.87844016938
Minibatch step 9/100: mean batch inertia: 88481.01554654661, ewa inertia: 88481.01554654661
Minibatch s

C:\Users\Brech\anaconda3\envs\chidi\lib\site-packages\sklearn\cluster\_kmeans.py:1043: UserWarning: MiniBatchKMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can prevent it by setting batch_size >= 3072 or by setting the environment variable OMP_NUM_THREADS=9
  warnings.warn(


Init 1/3 with method k-means++
Inertia for init 1/3: 287366911.0
Init 2/3 with method k-means++
Inertia for init 2/3: 289932061.0
Init 3/3 with method k-means++
Inertia for init 3/3: 287317337.0
Minibatch step 1/100: mean batch inertia: 139944.04212541887
Minibatch step 2/100: mean batch inertia: 91744.36843390563, ewa inertia: 91744.36843390563
Minibatch step 3/100: mean batch inertia: 91433.50770481514, ewa inertia: 91433.50770481514
Minibatch step 4/100: mean batch inertia: 89803.35960276482, ewa inertia: 89803.35960276482
Minibatch step 5/100: mean batch inertia: 90154.99817939915, ewa inertia: 90154.99817939915
Minibatch step 6/100: mean batch inertia: 88861.09752680836, ewa inertia: 88861.09752680836
Minibatch step 7/100: mean batch inertia: 88950.64544392459, ewa inertia: 88950.64544392459
Minibatch step 8/100: mean batch inertia: 89188.17020172012, ewa inertia: 89188.17020172012
Minibatch step 9/100: mean batch inertia: 89439.61809779269, ewa inertia: 89439.61809779269
Minibatc

C:\Users\Brech\anaconda3\envs\chidi\lib\site-packages\sklearn\cluster\_kmeans.py:1043: UserWarning: MiniBatchKMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can prevent it by setting batch_size >= 3072 or by setting the environment variable OMP_NUM_THREADS=9
  warnings.warn(


Init 1/3 with method k-means++
Inertia for init 1/3: 286654998.0
Init 2/3 with method k-means++
Inertia for init 2/3: 283456421.0
Init 3/3 with method k-means++
Inertia for init 3/3: 284281897.0
Minibatch step 1/100: mean batch inertia: 138044.95021541408
Minibatch step 2/100: mean batch inertia: 91990.08565533557, ewa inertia: 91990.08565533557
Minibatch step 3/100: mean batch inertia: 90403.22860998745, ewa inertia: 90403.22860998745
Minibatch step 4/100: mean batch inertia: 89063.39722097907, ewa inertia: 89063.39722097907
Minibatch step 5/100: mean batch inertia: 89716.45363432803, ewa inertia: 89716.45363432803
Minibatch step 6/100: mean batch inertia: 89607.08301292676, ewa inertia: 89607.08301292676
Minibatch step 7/100: mean batch inertia: 89106.75628316148, ewa inertia: 89106.75628316148
Minibatch step 8/100: mean batch inertia: 89408.50065959319, ewa inertia: 89408.50065959319
Minibatch step 9/100: mean batch inertia: 88462.41228102797, ewa inertia: 88462.41228102797
Minibatc

C:\Users\Brech\anaconda3\envs\chidi\lib\site-packages\sklearn\cluster\_kmeans.py:1043: UserWarning: MiniBatchKMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can prevent it by setting batch_size >= 3072 or by setting the environment variable OMP_NUM_THREADS=9
  warnings.warn(


Init 1/3 with method k-means++
Inertia for init 1/3: 289724466.0
Init 2/3 with method k-means++
Inertia for init 2/3: 287046198.0
Init 3/3 with method k-means++
Inertia for init 3/3: 284996233.0
Minibatch step 1/100: mean batch inertia: 136400.16515078986
Minibatch step 2/100: mean batch inertia: 91983.88024198968, ewa inertia: 91983.88024198968
Minibatch step 3/100: mean batch inertia: 91757.06807029838, ewa inertia: 91757.06807029838
Minibatch step 4/100: mean batch inertia: 90382.12741409143, ewa inertia: 90382.12741409143
Minibatch step 5/100: mean batch inertia: 90162.20349660741, ewa inertia: 90162.20349660741
Minibatch step 6/100: mean batch inertia: 90309.24807774293, ewa inertia: 90309.24807774293
Minibatch step 7/100: mean batch inertia: 89369.9913266299, ewa inertia: 89369.9913266299
Minibatch step 8/100: mean batch inertia: 89011.75826778282, ewa inertia: 89011.75826778282
Minibatch step 9/100: mean batch inertia: 89242.00407138029, ewa inertia: 89242.00407138029
Minibatch 

C:\Users\Brech\anaconda3\envs\chidi\lib\site-packages\sklearn\cluster\_kmeans.py:1043: UserWarning: MiniBatchKMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can prevent it by setting batch_size >= 3072 or by setting the environment variable OMP_NUM_THREADS=9
  warnings.warn(


Init 1/3 with method k-means++
Inertia for init 1/3: 285001565.0
Init 2/3 with method k-means++
Inertia for init 2/3: 287112735.0
Init 3/3 with method k-means++
Inertia for init 3/3: 282692725.0
Minibatch step 1/100: mean batch inertia: 135949.39636189563
Minibatch step 2/100: mean batch inertia: 91369.4288215292, ewa inertia: 91369.4288215292
Minibatch step 3/100: mean batch inertia: 89944.10025518796, ewa inertia: 89944.10025518796
Minibatch step 4/100: mean batch inertia: 89188.9994627241, ewa inertia: 89188.9994627241
Minibatch step 5/100: mean batch inertia: 89444.44985384705, ewa inertia: 89444.44985384705
Minibatch step 6/100: mean batch inertia: 88629.47709913262, ewa inertia: 88629.47709913262
Minibatch step 7/100: mean batch inertia: 88470.82996406217, ewa inertia: 88470.82996406217
Minibatch step 8/100: mean batch inertia: 87839.84712895681, ewa inertia: 87839.84712895681
Minibatch step 9/100: mean batch inertia: 88272.88262305764, ewa inertia: 88272.88262305764
Minibatch st

C:\Users\Brech\anaconda3\envs\chidi\lib\site-packages\sklearn\cluster\_kmeans.py:1043: UserWarning: MiniBatchKMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can prevent it by setting batch_size >= 3072 or by setting the environment variable OMP_NUM_THREADS=9
  warnings.warn(


Init 1/3 with method k-means++
Inertia for init 1/3: 284218770.0
Init 2/3 with method k-means++
Inertia for init 2/3: 290264917.0
Init 3/3 with method k-means++
Inertia for init 3/3: 287128838.0
Minibatch step 1/100: mean batch inertia: 136973.52848252753
Minibatch step 2/100: mean batch inertia: 91504.38423756321, ewa inertia: 91504.38423756321
Minibatch step 3/100: mean batch inertia: 90616.79544932967, ewa inertia: 90616.79544932967
Minibatch step 4/100: mean batch inertia: 90530.59760227968, ewa inertia: 90530.59760227968
Minibatch step 5/100: mean batch inertia: 89049.83309652048, ewa inertia: 89049.83309652048
Minibatch step 6/100: mean batch inertia: 89389.08295373282, ewa inertia: 89389.08295373282
Minibatch step 7/100: mean batch inertia: 90263.38013536009, ewa inertia: 90263.38013536009
Minibatch step 8/100: mean batch inertia: 89079.49723722303, ewa inertia: 89079.49723722303
Minibatch step 9/100: mean batch inertia: 88258.93729594268, ewa inertia: 88258.93729594268
Minibatc

C:\Users\Brech\anaconda3\envs\chidi\lib\site-packages\sklearn\cluster\_kmeans.py:1043: UserWarning: MiniBatchKMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can prevent it by setting batch_size >= 3072 or by setting the environment variable OMP_NUM_THREADS=9
  warnings.warn(


Init 1/3 with method k-means++
Inertia for init 1/3: 291378006.0
Init 2/3 with method k-means++
Inertia for init 2/3: 292012832.0
Init 3/3 with method k-means++
Inertia for init 3/3: 285588347.0
Minibatch step 1/100: mean batch inertia: 136216.87888942077
Minibatch step 2/100: mean batch inertia: 90940.9235330544, ewa inertia: 90940.9235330544
Minibatch step 3/100: mean batch inertia: 89874.40961309311, ewa inertia: 89874.40961309311
Minibatch step 4/100: mean batch inertia: 89332.09796469669, ewa inertia: 89332.09796469669
Minibatch step 5/100: mean batch inertia: 88986.4443708917, ewa inertia: 88986.4443708917
Minibatch step 6/100: mean batch inertia: 88664.00215713984, ewa inertia: 88664.00215713984
Minibatch step 7/100: mean batch inertia: 88530.58611893952, ewa inertia: 88530.58611893952
Minibatch step 8/100: mean batch inertia: 88977.73296063405, ewa inertia: 88977.73296063405
Minibatch step 9/100: mean batch inertia: 88665.48949702314, ewa inertia: 88665.48949702314
Minibatch st

C:\Users\Brech\anaconda3\envs\chidi\lib\site-packages\sklearn\cluster\_kmeans.py:1043: UserWarning: MiniBatchKMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can prevent it by setting batch_size >= 3072 or by setting the environment variable OMP_NUM_THREADS=9
  warnings.warn(


Init 1/3 with method k-means++
Inertia for init 1/3: 284515313.0
Init 2/3 with method k-means++
Inertia for init 2/3: 288693375.0
Init 3/3 with method k-means++
Inertia for init 3/3: 284312382.0
Minibatch step 1/100: mean batch inertia: 138951.20679751076
Minibatch step 2/100: mean batch inertia: 91914.11042847371, ewa inertia: 91914.11042847371
Minibatch step 3/100: mean batch inertia: 89782.06529601687, ewa inertia: 89782.06529601687
Minibatch step 4/100: mean batch inertia: 89454.8778983082, ewa inertia: 89454.8778983082
Minibatch step 5/100: mean batch inertia: 89227.42873290119, ewa inertia: 89227.42873290119
Minibatch step 6/100: mean batch inertia: 90076.80098852377, ewa inertia: 90076.80098852377
Minibatch step 7/100: mean batch inertia: 88821.30681513858, ewa inertia: 88821.30681513858
Minibatch step 8/100: mean batch inertia: 88367.68750590215, ewa inertia: 88367.68750590215
Minibatch step 9/100: mean batch inertia: 88516.90955420045, ewa inertia: 88516.90955420045
Minibatch 

C:\Users\Brech\anaconda3\envs\chidi\lib\site-packages\sklearn\cluster\_kmeans.py:1043: UserWarning: MiniBatchKMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can prevent it by setting batch_size >= 3072 or by setting the environment variable OMP_NUM_THREADS=9
  warnings.warn(


Init 1/3 with method k-means++
Inertia for init 1/3: 290026913.0
Init 2/3 with method k-means++
Inertia for init 2/3: 289403253.0
Init 3/3 with method k-means++
Inertia for init 3/3: 288397767.0
Minibatch step 1/100: mean batch inertia: 139567.68884633796
Minibatch step 2/100: mean batch inertia: 91458.42338827484, ewa inertia: 91458.42338827484
Minibatch step 3/100: mean batch inertia: 89745.25204021663, ewa inertia: 89745.25204021663
Minibatch step 4/100: mean batch inertia: 89513.92028581498, ewa inertia: 89513.92028581498
Minibatch step 5/100: mean batch inertia: 89371.16473354516, ewa inertia: 89371.16473354516
Minibatch step 6/100: mean batch inertia: 89655.70240116859, ewa inertia: 89655.70240116859
Minibatch step 7/100: mean batch inertia: 88287.18299807311, ewa inertia: 88287.18299807311
Minibatch step 8/100: mean batch inertia: 88421.81604494035, ewa inertia: 88421.81604494035
Minibatch step 9/100: mean batch inertia: 88263.17581181342, ewa inertia: 88263.17581181342
Minibatc

C:\Users\Brech\anaconda3\envs\chidi\lib\site-packages\sklearn\cluster\_kmeans.py:1043: UserWarning: MiniBatchKMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can prevent it by setting batch_size >= 3072 or by setting the environment variable OMP_NUM_THREADS=9
  warnings.warn(


Init 1/3 with method k-means++
Inertia for init 1/3: 291209842.0
Init 2/3 with method k-means++
Inertia for init 2/3: 290578165.0
Init 3/3 with method k-means++
Inertia for init 3/3: 286756676.0
Minibatch step 1/100: mean batch inertia: 136860.92101483964
Minibatch step 2/100: mean batch inertia: 90905.13090136985, ewa inertia: 90905.13090136985
Minibatch step 3/100: mean batch inertia: 90207.23600475634, ewa inertia: 90207.23600475634
Minibatch step 4/100: mean batch inertia: 88983.0979907, ewa inertia: 88983.0979907
Minibatch step 5/100: mean batch inertia: 88463.5206334856, ewa inertia: 88463.5206334856
Minibatch step 6/100: mean batch inertia: 88240.42573966074, ewa inertia: 88240.42573966074
Minibatch step 7/100: mean batch inertia: 88736.16191746548, ewa inertia: 88736.16191746548
Minibatch step 8/100: mean batch inertia: 88078.33504280835, ewa inertia: 88078.33504280835
Minibatch step 9/100: mean batch inertia: 88120.18336865648, ewa inertia: 88120.18336865648
Minibatch step 10/

In [154]:
sample_feats_df = pd.DataFrame(sample_feats)

In [151]:
sample_feats

[array([0.06378738, 0.04186047, 0.03654485, 0.03189369, 0.03853821,
        0.01960133, 0.0372093 , 0.04418605, 0.0282392 , 0.03920266,
        0.01860465, 0.03853821, 0.02425249, 0.03289037, 0.02757475,
        0.03222591, 0.02591362, 0.04186047, 0.03189369, 0.02491694,
        0.02093023, 0.03122924, 0.03654485, 0.01594684, 0.04019934,
        0.02259136, 0.05282392, 0.03421927, 0.02425249, 0.04152824]),
 array([0.04551724, 0.03517241, 0.02275862, 0.04448276, 0.04758621,
        0.02551724, 0.02793103, 0.06689655, 0.01689655, 0.03206897,
        0.0337931 , 0.02551724, 0.03206897, 0.02275862, 0.04551724,
        0.04310345, 0.02689655, 0.02517241, 0.01896552, 0.02206897,
        0.03758621, 0.02655172, 0.04241379, 0.03413793, 0.03137931,
        0.03965517, 0.02862069, 0.02655172, 0.04172414, 0.03068966]),
 array([0.01561416, 0.02741152, 0.04510756, 0.02151284, 0.05898681,
        0.01873699, 0.04024983, 0.03990285, 0.04510756, 0.02637058,
        0.03469813, 0.04614851, 0.04649549, 

In [155]:
sample_feats_df

0         1         2         3         4         5         6   \
0   0.063787  0.041860  0.036545  0.031894  0.038538  0.019601  0.037209   
1   0.045517  0.035172  0.022759  0.044483  0.047586  0.025517  0.027931   
2   0.015614  0.027412  0.045108  0.021513  0.058987  0.018737  0.040250   
3   0.030602  0.021761  0.017681  0.032982  0.030942  0.028222  0.034682   
4   0.045194  0.024283  0.023946  0.048229  0.030691  0.031703  0.035750   
5   0.030753  0.052522  0.022460  0.034900  0.044229  0.022460  0.024188   
6   0.029282  0.040518  0.031324  0.028941  0.040518  0.032346  0.059925   
7   0.035423  0.031173  0.021962  0.029756  0.024088  0.024442  0.040383   
8   0.023305  0.079802  0.026483  0.021540  0.027542  0.032486  0.033898   
9   0.025321  0.079529  0.028174  0.053138  0.031027  0.024251  0.037447   
10  0.028690  0.038714  0.029036  0.050467  0.035949  0.026962  0.032492   
11  0.031962  0.041142  0.039442  0.025842  0.030602  0.024141  0.037742   
12  0.016518  0.061597  0.028906  0.039573  0.039229  0.043014  0.030970   
13  0.025478  0.042575  0.035535  0.020114  0.030841  0.046597  0.026483   
14  0.025948  0.032269  0.025948  0.036261  0.031936  0.032934  0.044245   
15  0.008348  0.051826  0.037913  0.042783  0.029913  0.031652  0.018435   
16  0.015753  0.031835  0.040696  0.019035  0.031178  0.020676  0.017722   
17  0.040027  0.047151  0.024763  0.035617  0.024423  0.054613  0.029851   
18  0.045224  0.030712  0.029362  0.054674  0.025650  0.017212  0.019575   
19  0.020681  0.052498  0.027044  0.041362  0.061406  0.016545  0.025135   
20  0.026368  0.033289  0.052406  0.030323  0.029993  0.032630  0.044496   
21  0.022696  0.037263  0.038957  0.024051  0.032520  0.032520  0.043360   
22  0.022878  0.035477  0.042440  0.026857  0.025862  0.044430  0.030172   
23  0.027293  0.034592  0.073310  0.036814  0.037131  0.019676  0.024437   
24  0.025834  0.037316  0.014352  0.042698  0.044133  0.042339  0.033010   
25  0.021565  0.034783  0.044870  0.048000  0.018087  0.032000  0.030609   
26  0.045135  0.029756  0.009027  0.049482  0.040455  0.028419  0.035105   
27  0.013252  0.029901  0.025824  0.045532  0.033639  0.039755  0.048590   
28  0.038369  0.039054  0.023638  0.032545  0.031175  0.060980  0.030490   
29  0.021850  0.050188  0.031751  0.046091  0.030386  0.020826  0.018095   
30  0.028075  0.039773  0.052139  0.044452  0.053810  0.028409  0.036765   
31  0.042112  0.022393  0.027741  0.034091  0.027741  0.020722  0.033088   
32  0.027901  0.040830  0.041851  0.027901  0.019394  0.036747  0.033345   
33  0.029795  0.018932  0.021726  0.037554  0.031037  0.067349  0.022657   
34  0.030500  0.027255  0.034718  0.023037  0.032446  0.044452  0.021415   
35  0.023103  0.023448  0.049655  0.040690  0.032759  0.025862  0.036552   
36  0.029207  0.029207  0.032526  0.034185  0.029871  0.021573  0.024560   
37  0.051105  0.018992  0.026588  0.038674  0.060083  0.025207  0.026934   
38  0.027721  0.026694  0.023614  0.041410  0.045517  0.057495  0.023272   
39  0.059399  0.023962  0.024975  0.015862  0.035775  0.043874  0.022950   
40  0.018266  0.029558  0.053803  0.030223  0.050149  0.038858  0.037861   
41  0.049482  0.035105  0.016717  0.030759  0.037780  0.038449  0.040789   
42  0.025519  0.040150  0.038789  0.049677  0.040150  0.029602  0.020415   
43  0.043424  0.032725  0.012901  0.032410  0.020138  0.025173  0.032725   
44  0.042568  0.027806  0.029866  0.018881  0.019911  0.041881  0.048747   
45  0.017952  0.037566  0.033910  0.036237  0.020612  0.032247  0.023271   
46  0.030666  0.049700  0.033134  0.036306  0.035249  0.032781  0.047938   
47  0.028689  0.034836  0.058060  0.024932  0.052254  0.024590  0.045423   
48  0.029620  0.024907  0.026590  0.049142  0.023225  0.034332  0.049815   
49  0.022356  0.017684  0.034368  0.022022  0.032032  0.034701  0.031698   

          7         8         9   ...        20        21        22        23  \
0   0.044186  0.028239  0.039203  ...  0.020930  0.0

Another thing I'll be doing to reduce the abuse on my CPU, is saving this sample in a separate csv-file in the data-folder. This way, I won't have to re-execute this operation over and over again.

In [156]:
sample_feats_df.to_csv('data/brecht-sample.csv')

In [162]:
from sklearn.model_selection import train_test_split

X = pd.read_csv("data/brecht-sample.csv")
Y = sample["choice"]

X_train, X_test, y_train, y_test = train_test_split(X, Y, test_size=0.4, random_state=0)

#### The Decision Tree Model

In [166]:
from sklearn.tree import DecisionTreeClassifier
dec_tree = DecisionTreeClassifier()
dec_tree = dec_tree.fit(X_train, y_train)

In [171]:
dec_tree.score(X_test, y_test)

0.65

#### The Logistic Regression Model

In [168]:
from sklearn.linear_model import LogisticRegressionCV
log_reg = LogisticRegressionCV()
log_reg = log_reg.fit(X_train, y_train)

C:\Users\Brech\anaconda3\envs\chidi\lib\site-packages\sklearn\linear_model\_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
C:\Users\Brech\anaconda3\envs\chidi\lib\site-packages\sklearn\linear_model\_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logisti

In [170]:
log_reg.score(X_test, y_test)

0.75

#### The "Perceptotron" (non-deep neural network)

### Deep-learning neural networks

Any machine learning model is considered deep learning if it has at least three layers total. This includes input layers, output layers and the hidden layers in between.

When it comes to computer vision, neural networks have become the preferred option. I'll be making several neural networks with varying numbers of hidden layers. This is to test the impact an increasing number of layers has on the speed and performance of the model.

It wouldn't be an analysis if I didn't have an initial hypothesis: I'm going to start off from the viewpoint that the performance of our model will be directly proportional to the total number of hidden layers, while our speed will be inversely proportional to it. In other words: the more layers, the better our model performs, but the longer it will take to make a prediction. 

In [172]:
from sklearn.model_selection import train_test_split

#Well, what are we waiting for? Let's load in our data and get analyzing!
sample = os.listdir("data/images/3 De nayer (Loop piste)/partition 1")
sample = data.loc[data["image"].isin(sample)].head(50)

In [173]:
#Our X will contain the pixels of our loaded images, while our Y contains all of the classifiers
X = np.array([cv.imread("/".join(['data','images','3 De nayer (Loop piste)','partition 1',img])) for img in sample["image"]])
Y = np.array(sample["choice"])

X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.4, random_state=0)

Our training set consists of 30 images, each of which consist of 480 x 640 BRG-pixels.

In [187]:
X_train.shape

(30, 480, 640, 3)

In [188]:
len(Y_train)

30

In [191]:
Y_train

array([2., 2., 2., 2., 0., 2., 2., 1., 2., 2., 0., 2., 1., 1., 0., 2., 2.,
       2., 0., 2., 2., 2., 2., 2., 1., 2., 0., 2., 1., 2.])

In [192]:
X_train = X_train / 255.0

In [195]:
X_test = X_test / 255.0

#### The models I made

* First Model: Includes only a Flatten-layer and a Dense-layer, the latter consisting of 3 neurons (as many as there are classes) and activated by the softmax-function.
* Second Model: Consists of 6 layers, including input and output, for a total of 4 hidden layers.
* Third Model: Consists of 8 layers, including input and output, for a total of 6 hidden layers.

In [196]:
first_model = tf.keras.models.Sequential([
    tf.keras.layers.Flatten(input_shape=(480,640,3)),
    tf.keras.layers.Dense(3, activation='softmax')
])

second_model = tf.keras.models.Sequential([
    tf.keras.layers.Flatten(input_shape=(480,640,3)),
    tf.keras.layers.Dense(128, activation='relu'),
    tf.keras.layers.Dense(32, activation='relu'),
    tf.keras.layers.Dense(3, activation='softmax')
])

third_model = tf.keras.models.Sequential([
    tf.keras.layers.Flatten(input_shape=(480,640,3)),
    tf.keras.layers.Dense(256, activation='relu'),
    tf.keras.layers.Dense(256, activation='selu'),
    tf.keras.layers.Dense(128, activation='relu'),
    tf.keras.layers.Dense(128, activation='selu'),
    tf.keras.layers.Dense(64, activation='relu'),
    tf.keras.layers.Dense(32, activation='relu'),
    tf.keras.layers.Dense(3, activation='softmax')
])

In essence, my first model is nothing more than essentially our Logistic Regression-model from earlier converted into a neural network.

In [197]:
first_model.compile(optimizer=tf.keras.optimizers.Adam(learning_rate=3e-4),
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

first_model.summary()

Model: "sequential_6"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 flatten_6 (Flatten)         (None, 921600)            0         
                                                                 
 dense_22 (Dense)            (None, 3)                 2764803   
                                                                 
Total params: 2,764,803
Trainable params: 2,764,803
Non-trainable params: 0
_________________________________________________________________


In [210]:
first_model.fit(X_train, Y_train, epochs=25, batch_size=10, validation_split=0.2)

Epoch 1/25
3/3 [==============================] - 0s 144ms/step - loss: 2.3445 - accuracy: 0.7083 - val_loss: 14.2155 - val_accuracy: 0.5000
Epoch 2/25
3/3 [==============================] - 0s 130ms/step - loss: 9.3181 - accuracy: 0.7083 - val_loss: 8.3855 - val_accuracy: 0.5000
Epoch 3/25
3/3 [==============================] - 0s 99ms/step - loss: 5.0289 - accuracy: 0.8333 - val_loss: 19.7735 - val_accuracy: 0.1667
Epoch 4/25
3/3 [==============================] - 0s 96ms/step - loss: 5.0639 - accuracy: 0.5833 - val_loss: 12.5198 - val_accuracy: 0.5000
Epoch 5/25
3/3 [==============================] - 0s 108ms/step - loss: 3.8144 - accuracy: 0.6250 - val_loss: 12.8609 - val_accuracy: 0.5000
Epoch 6/25
3/3 [==============================] - 0s 111ms/step - loss: 8.7390 - accuracy: 0.6250 - val_loss: 8.6029 - val_accuracy: 0.3333
Epoch 7/25
3/3 [==============================] - 0s 110ms/step - loss: 9.2138 - accuracy: 0.5000 - val_loss: 21.2543 - val_accuracy: 0.5000
Epoch 8/25
3/3 [=

In [211]:
first_model.evaluate(X_test, Y_test)

1/1 [==============================] - 0s 70ms/step - loss: 5.4937 - accuracy: 0.8500


[5.493687629699707, 0.8500000238418579]

Our first model takes 74 milliseconds to make a prediction, with an accuracy of 80%. Performance-wise, it is better than even our Logistic Regression-model earlier. Considering this is supposed to be the "worst" deep learning model, I have high hopes for the rest of the models.

In [200]:
second_model.compile(optimizer=tf.keras.optimizers.Adam(learning_rate=3e-4),
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

second_model.summary()

Model: "sequential_7"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 flatten_7 (Flatten)         (None, 921600)            0         
                                                                 
 dense_23 (Dense)            (None, 128)               117964928 
                                                                 
 dense_24 (Dense)            (None, 32)                4128      
                                                                 
 dense_25 (Dense)            (None, 3)                 99        
                                                                 
Total params: 117,969,155
Trainable params: 117,969,155
Non-trainable params: 0
_________________________________________________________________


In [212]:
second_model.fit(X_train, Y_train, epochs=25, batch_size=10, validation_split=0.2)

Epoch 1/25
3/3 [==============================] - 4s 1s/step - loss: 0.7012 - accuracy: 0.7500 - val_loss: 1.3715 - val_accuracy: 0.5000
Epoch 2/25
3/3 [==============================] - 3s 1s/step - loss: 2.2412 - accuracy: 0.2083 - val_loss: 1.7491 - val_accuracy: 0.5000
Epoch 3/25
3/3 [==============================] - 4s 1s/step - loss: 2.2687 - accuracy: 0.5000 - val_loss: 5.1659 - val_accuracy: 0.5000
Epoch 4/25
3/3 [==============================] - 4s 1s/step - loss: 3.0736 - accuracy: 0.7083 - val_loss: 9.7794 - val_accuracy: 0.5000
Epoch 5/25
3/3 [==============================] - 4s 1s/step - loss: 3.9725 - accuracy: 0.7500 - val_loss: 7.4839 - val_accuracy: 0.5000
Epoch 6/25
3/3 [==============================] - 4s 1s/step - loss: 2.5652 - accuracy: 0.7083 - val_loss: 3.8156 - val_accuracy: 0.5000
Epoch 7/25
3/3 [==============================] - 4s 1s/step - loss: 2.3077 - accuracy: 0.5417 - val_loss: 6.3970 - val_accuracy: 0.5000
Epoch 8/25
3/3 [=========================

Training this model took considerably longer than our first model, with every epoch taking a handful of seconds to complete. The longest epoch in our last model lasted about one second.

In [213]:
second_model.evaluate(X_test, Y_test)

1/1 [==============================] - 0s 121ms/step - loss: 2.0548 - accuracy: 0.7500


[2.054837465286255, 0.75]

Speed seems to have gone down to 121 milliseconds per prediction made, but the model still has an accuracy 5% higher than our first neural network, and 10% higher than our Logistic Regression-model.

It does, however, depend on the data used. In a previous attempt, before the pixel-values of our images were converted to values between 0 and 1, our second model actually performed *worse* than not just our first model, but also every single one of our non-deep learning models. The performance of a neural network, therefore, is not just dependent on the structure of its layers, but also on the data fed into its inputs.

In [203]:
third_model.compile(optimizer=tf.keras.optimizers.Adam(learning_rate=3e-4),
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

third_model.summary()

Model: "sequential_8"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 flatten_8 (Flatten)         (None, 921600)            0         
                                                                 
 dense_26 (Dense)            (None, 256)               235929856 
                                                                 
 dense_27 (Dense)            (None, 256)               65792     
                                                                 
 dense_28 (Dense)            (None, 128)               32896     
                                                                 
 dense_29 (Dense)            (None, 128)               16512     
                                                                 
 dense_30 (Dense)            (None, 64)                8256      
                                                                 
 dense_31 (Dense)            (None, 32)               

In [207]:
third_model.fit(X_train, Y_train, epochs=25, batch_size=10, validation_split=0.2)

Epoch 1/25
3/3 [==============================] - 8s 2s/step - loss: 8.7798 - accuracy: 0.6250 - val_loss: 8.7169 - val_accuracy: 0.5000
Epoch 2/25
3/3 [==============================] - 8s 3s/step - loss: 3.9329 - accuracy: 0.7083 - val_loss: 3.9863 - val_accuracy: 0.5000
Epoch 3/25
3/3 [==============================] - 8s 3s/step - loss: 2.3695 - accuracy: 0.6667 - val_loss: 14.9114 - val_accuracy: 0.1667
Epoch 4/25
3/3 [==============================] - 8s 3s/step - loss: 5.5436 - accuracy: 0.5833 - val_loss: 19.3375 - val_accuracy: 0.1667
Epoch 5/25
3/3 [==============================] - 8s 3s/step - loss: 8.3505 - accuracy: 0.4583 - val_loss: 23.5491 - val_accuracy: 0.5000
Epoch 6/25
3/3 [==============================] - 8s 3s/step - loss: 15.0465 - accuracy: 0.7083 - val_loss: 39.3249 - val_accuracy: 0.5000
Epoch 7/25
3/3 [==============================] - 8s 3s/step - loss: 20.6045 - accuracy: 0.3750 - val_loss: 30.2747 - val_accuracy: 0.5000
Epoch 8/25
3/3 [==================

In [208]:
third_model.evaluate(X_test, Y_test)

1/1 [==============================] - 0s 201ms/step - loss: 8.0819 - accuracy: 0.1500


[8.081899642944336, 0.15000000596046448]

Every prediction took about 169 milliseconds, and our accuracy has shot up to above 75%.

### Conclusions

* The way the data fed into a neural network is structured plays a big role in determining the overall accuracy of a neural network when classifying images.